bootstrap/pkg/kfapp/coordinator/coordinator.go

```diff
@@ -241,205 +251,393 @@ func NewKfApp(options map[string]interface{}) (kftypes.KfApp, error) {
+
 	}
 
-	kfDef.Name = appName
-	kfDef.Spec.AppDir = appDir
-	kfDef.Spec.Platform = platform
-	kfDef.Namespace = namespace
-	kfDef.Spec.Version = version
-	kfDef.Spec.Repo = path.Join(cacheDir, kftypes.KubeflowRepo)
-	kfDef.Spec.Project = options[string(kftypes.PROJECT)].(string)
-	kfDef.Spec.SkipInitProject = options[string(kftypes.SKIP_INIT_GCP_PROJECT)].(bool)
-	kfDef.Spec.UseBasicAuth = useBasicAuth
-	kfDef.Spec.UseIstio = useIstio
-	kfDef.Spec.PackageManager = packageManager
-	pApp := GetKfApp(kfDef, nil)
-	return pApp, nil
+	return kfDef, nil
 }
 
-// unmarshalAppYaml is a local function to marshal the contents of app.yaml into
-// the KfDef type
-func unmarshalAppYaml(cfgfile string, kfdef *kfdefsv2.KfDef) error {
-	if _, err := os.Stat(cfgfile); err == nil {
-		log.Infof("reading from %v", cfgfile)
-		buf, bufErr := ioutil.ReadFile(cfgfile)
-		if bufErr != nil {
-			return &kfapis.KfError{
-				Code:    int(kfapis.INVALID_ARGUMENT),
-				Message: fmt.Sprintf("couldn't read %v. Error: %v", cfgfile, bufErr),
-			}
-		}
-		err := yaml.Unmarshal(buf, kfdef)
-		if err != nil {
-			return &kfapis.KfError{
-				Code:    int(kfapis.INTERNAL_ERROR),
-				Message: fmt.Sprintf("could not unmarshal %v. Error: %v", cfgfile, err),
-			}
+// CreateKfAppDir will create the application directory and persist
+// the KfDef to it as app.yaml.
+// Returns an error if the app.yaml file already exists
+// Returns path to the app.yaml file.
+func CreateKfAppCfgFile(d *kfdefsv2.KfDef) (string, error) {
+	if _, err := os.Stat(d.Spec.AppDir); os.IsNotExist(err) {
+		log.Infof("Creating directory %v", d.Spec.AppDir)
+		appdirErr := os.MkdirAll(d.Spec.AppDir, os.ModePerm)
+		if appdirErr != nil {
+			log.Errorf("couldn't create directory %v Error %v", d.Spec.AppDir, appdirErr)
+			return "", appdirErr
 		}
+	} else {
+		log.Infof("App directory %v already exists", d.Spec.AppDir)
 	}
-	return nil
+
+	// Rewrite app.yaml
+	cfgFilePath := filepath.Join(d.Spec.AppDir, kftypesv2.KfConfigFile)
+
+	if _, err := os.Stat(cfgFilePath); err == nil {
+		log.Errorf("%v already exists", cfgFilePath)
+		return cfgFilePath, fmt.Errorf("%v already exists", cfgFilePath)
+	}
+	log.Infof("Writing KfDef to %v", cfgFilePath)
+	cfgFilePathErr := d.WriteToFile(cfgFilePath)
+	return cfgFilePath, cfgFilePathErr
 }
 
-// LoadKfApp is called from subcommands Apply, Delete, Generate and assumes the existence of an app.yaml
-// file which was created by the Init subcommand. It sets options needed by these subcommands
-func LoadKfApp(options map[string]interface{}) (kftypes.KfApp, error) {
-	appDir, err := os.Getwd()
+// NewKfApp is called from the Init subcommand and will create a directory based on
+// the path/name argument given to the Init subcommand
+func NewKfApp(options map[string]interface{}) (kftypes.KfApp, error) {
+	kfDef, err := CreateKfDefFromOptions(options)
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = backfillKfDefFromInitOptions(kfDef, options)
```

Comment:
I can see that you need to back fill generate option when calling LoadKfApp.
But why do we need this here? Can't the work here be done in CreateKfDefFromOptions(options) above?

Reply:
Done.